examples/custom-environment/lib.rs

```diff
@@ -0,0 +1,132 @@
+    const MAX_EVENT_TOPICS: usize =
+        <ink::env::DefaultEnvironment as Environment>::MAX_EVENT_TOPICS + 1;
+
+    type AccountId = <ink::env::DefaultEnvironment as Environment>::AccountId;
+    type Balance = <ink::env::DefaultEnvironment as Environment>::Balance;
+    type Hash = <ink::env::DefaultEnvironment as Environment>::Hash;
+    type BlockNumber = <ink::env::DefaultEnvironment as Environment>::BlockNumber;
+    type Timestamp = <ink::env::DefaultEnvironment as Environment>::Timestamp;
+
+    type ChainExtension = <ink::env::DefaultEnvironment as Environment>::ChainExtension;
+}
+
+#[ink::contract(env = crate::EnvironmentWithManyTopics)]
+mod runtime_call {
+    /// Trivial contract with a single message that emits an event with many topics.
+    #[ink(storage)]
+    #[derive(Default)]
+    pub struct Topicer;
+
+    /// An event that would be forbidden in the default environment, but is completely valid in
+    /// our custom one.
+    #[ink(event)]
+    #[derive(Default)]
+    pub struct TopicedEvent {
+        #[ink(topic)]
+        first_topic: Balance,
+        #[ink(topic)]
+        second_topic: Balance,
+        #[ink(topic)]
+        third_topic: Balance,
+        #[ink(topic)]
+        fourth_topic: Balance,
+        #[ink(topic)]
+        fifth_topic: Balance,
+    }
+
+    impl Topicer {
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Self {}
+        }
+
+        /// Emit an event with many topics.
+        #[ink(message)]
+        pub fn trigger(&mut self) {
+            self.env().emit_event(TopicedEvent::default());
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+
+        #[ink::test]
+        fn emits_event_with_many_topics() {
+            let mut contract = Topicer::new();
+            contract.trigger();
+
+            let emitted_events = ink::env::test::recorded_events().collect::<Vec<_>>();
+            assert_eq!(1, emitted_events.len());
+
+            let _ = <<Topicer as ink::reflect::ContractEventBase>::Type as scale::Decode>::decode(&mut &emitted_events[0].data[..])
+                .expect("encountered invalid contract event data buffer");
+        }
+    }
+
+    #[cfg(all(test, feature = "e2e-tests"))]
+    mod e2e_tests {
+        use super::*;
+        use crate::EnvironmentWithManyTopics;
+        use ink_e2e::MessageBuilder;
+
+        type E2EResult<T> = Result<T, Box<dyn std::error::Error>>;
+
+        #[ink_e2e::test(environment = crate::EnvironmentWithManyTopics)]
+        #[ignore = "Requires that the pallet contract is configured with a schedule allowing for \
+            more event topics. For example:\
+            ```rust
+            pub Schedule: pallet_contracts::Schedule<Runtime> = pallet_contracts::Schedule::<Runtime> {
+        		limits: pallet_contracts::Limits {
+		    	    event_topics: 6,
+			        ..Default::default()
+		        },
+           		..Default::default()
+	        };
```

Comment:
```suggestion
            pub Schedule: pallet_contracts::Schedule<Runtime> = pallet_contracts::Schedule::<Runtime> {
                limits: pallet_contracts::Limits {
                    event_topics: 6,
                    ..Default::default()
                },
                ..Default::default()
            };
```

spaces pls, lol

Reply:
This also doesn't look formatted nicely when run. We may want to add some line breaks and get rid of the `rust` block

Reply:
Similar to the other PR, we should feature flag this and a test showing what an error would look like

Reply:
tabs got inserted somehow instead of spaces; again - how would you call a feature?

Reply:
added `permissive-node` feature